In [1]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/raw")        # notebook is in notebooks/, so go up one level

tel = pd.read_csv(DATA_DIR / "PdM_telemetry.csv")
print(tel["datetime"].head(3))        # what does the raw value look like?
print(tel.dtypes)                     # BEFORE: datetime is text

tel["datetime"] = pd.to_datetime(tel["datetime"])
print(tel.dtypes)                     # AFTER: datetime64
print(tel["datetime"].min(), "→", tel["datetime"].max())

0    2015-01-01 06:00:00
1    2015-01-01 07:00:00
2    2015-01-01 08:00:00
Name: datetime, dtype: str
datetime         str
machineID      int64
volt         float64
rotate       float64
pressure     float64
vibration    float64
dtype: object
datetime     datetime64[us]
machineID             int64
volt                float64
rotate              float64
pressure            float64
vibration           float64
dtype: object
2015-01-01 06:00:00 → 2016-01-01 06:00:00


In [2]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/raw")        # notebook is in notebooks/, so go up one level

failures = pd.read_csv(DATA_DIR / "PdM_failures.csv")
print(failures["datetime"].head(3))        # what does the raw value look like?
print(failures.dtypes)                     # BEFORE: datetime is text

failures["datetime"] = pd.to_datetime(failures["datetime"])
print(failures.dtypes)                     # AFTER: datetime64
print(failures["datetime"].min(), "→", failures["datetime"].max())

0    2015-01-05 06:00:00
1    2015-03-06 06:00:00
2    2015-04-20 06:00:00
Name: datetime, dtype: str
datetime       str
machineID    int64
failure        str
dtype: object
datetime     datetime64[us]
machineID             int64
failure                 str
dtype: object
2015-01-02 03:00:00 → 2015-12-31 06:00:00


In [3]:
from pathlib import Path

In [4]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/raw")

FILES = {
    "telemetry": "PdM_telemetry.csv",
    "errors":    "PdM_errors.csv",
    "maint":     "PdM_maint.csv",
    "failures":  "PdM_failures.csv",
    "machines":  "PdM_machines.csv",
}

tables = {}
for name, file in FILES.items():
    df = pd.read_csv(DATA_DIR / file)

    if "datetime" in df.columns:
        df["datetime"] = pd.to_datetime(df["datetime"])   # only for tables that have dates
        date_info = str(df["datetime"].dtype)
    else:
        date_info = "no datetime column"                    # machines lands here

    tables[name] = df                                       # store every table, dates or not
    print(f"{name:<10} {str(df.shape):<14} {date_info}")

telemetry  (876100, 6)    datetime64[us]
errors     (3919, 3)      datetime64[us]
maint      (3286, 3)      datetime64[us]
failures   (761, 3)       datetime64[us]
machines   (100, 3)       no datetime column


In [5]:
from pathlib import Path
import pandas as pd
DATA_DIR = Path("../data/raw")        # notebook is in notebooks/, so go up one level

maint = pd.read_csv(DATA_DIR / "PdM_maint.csv")
print(maint["datetime"].head(3))        # what does the raw value look like?
print(maint.dtypes)                     # BEFORE: datetime is text

maint["datetime"] = pd.to_datetime(maint["datetime"])
print(maint.dtypes)                     # AFTER: datetime64
print(maint["datetime"].min(), "→", maint["datetime"].max())

0    2014-06-01 06:00:00
1    2014-07-16 06:00:00
2    2014-07-31 06:00:00
Name: datetime, dtype: str
datetime       str
machineID    int64
comp           str
dtype: object
datetime     datetime64[us]
machineID             int64
comp                    str
dtype: object
2014-06-01 06:00:00 → 2016-01-01 06:00:00


In [6]:
from pathlib import Path
import pandas as pd
DATA_DIR = Path("../data/raw")        # notebook is in notebooks/, so go up one level

telemetry = pd.read_csv(DATA_DIR / "PdM_maint.csv")
print(telemetry["datetime"].head(3))        # what does the raw value look like?
print(telemetry.dtypes)                     # BEFORE: datetime is text

telemetry["datetime"] = pd.to_datetime(telemetry["datetime"])
print(telemetry.dtypes)                     # AFTER: datetime64
print(telemetry["datetime"].min(), "→", telemetry["datetime"].max())

0    2014-06-01 06:00:00
1    2014-07-16 06:00:00
2    2014-07-31 06:00:00
Name: datetime, dtype: str
datetime       str
machineID    int64
comp           str
dtype: object
datetime     datetime64[us]
machineID             int64
comp                    str
dtype: object
2014-06-01 06:00:00 → 2016-01-01 06:00:00


In [7]:
from pathlib import Path
import pandas as pd

In [8]:
from pathlib import Path
import pandas as pd
DATA_DIR = Path("../data/raw")
telemetry = pd.read_csv(DATA_DIR / "PdM_maint.csv")
print(telemetry["datetime"].head(3))        # what does the raw value look like?

0    2014-06-01 06:00:00
1    2014-07-16 06:00:00
2    2014-07-31 06:00:00
Name: datetime, dtype: str


In [9]:
for name, df in tables.items():
    if "datetime" in df.columns:
        print(f"{name:<10}", df["datetime"].dtype)

telemetry  datetime64[us]
errors     datetime64[us]
maint      datetime64[us]
failures   datetime64[us]


In [10]:
def first_look(df, name):
    print("=" * 60)
    print(name, df.shape)     # shape: a value, so no ()
    print(df.head(3))         # head: an action, so ()
    print(df.tail(3))

first_look(tables["failures"], "failures")

failures (761, 3)
             datetime  machineID failure
0 2015-01-05 06:00:00          1   comp4
1 2015-03-06 06:00:00          1   comp1
2 2015-04-20 06:00:00          1   comp2
               datetime  machineID failure
758 2015-02-12 06:00:00        100   comp1
759 2015-09-10 06:00:00        100   comp1
760 2015-12-09 06:00:00        100   comp2


In [12]:
def first_look(df, name):   # def = "define a recipe" called first_look
                            # (df, name) = two blanks to fill in each time you use it:
                            #    df   → which table to inspect
                            #    name → a label to print, like "failures"
                            # the colon means "the recipe's steps come next"

    print("=" * 60)         # "=" repeated 60 times: a divider line between tables
    print(name, df.shape)   # the label, then (rows, columns); shape is a value, so no ()
    print(df.head(3))       # the first 3 rows
    print(df.tail(3))       # the last 3 rows
    print()                 # an empty line for better readability between tables

In [13]:
def first_look(df, name):
    """Print a quick health check of one table."""
    print("=" * 60)
    print(f"{name.upper()}   shape: {df.shape}")

    print("\n--- first and last rows ---")
    print(df.head(3))
    print(df.tail(3))

    print("\n--- column types ---")                       # step 3
    print(df.dtypes)

    print("\n--- missing values per column ---")          # step 4
    print(df.isna().sum())
    print("duplicate rows:", df.duplicated().sum())

    if "datetime" in df.columns:                           # step 5: only tables with dates
        print("\n--- time range ---")
        print(df["datetime"].min(), "→", df["datetime"].max())

    if "machineID" in df.columns:                          # step 6
        print("\nunique machines:", df["machineID"].nunique())

    if "machineID" in df.columns and "datetime" in df.columns:
        dup_keys = df.duplicated(subset=["machineID", "datetime"]).sum()
        print("same machine + same time rows:", dup_keys)

    text_cols = [c for c in df.columns                     # step 7: find the text columns
                 if pd.api.types.is_string_dtype(df[c])]
    for col in text_cols:
        print(f"\n--- value counts: {col} ---")
        print(df[col].value_counts())

    num = df.select_dtypes("number").drop(columns="machineID", errors="ignore")
    if num.shape[1] > 0:                                   # step 8: only if number columns remain
        print("\n--- number summary ---")
        print(num.describe().round(2))

In [14]:
for name, df in tables.items():
    first_look(df, name)

TELEMETRY   shape: (876100, 6)

--- first and last rows ---
             datetime  machineID        volt      rotate    pressure  \
0 2015-01-01 06:00:00          1  176.217853  418.504078  113.077935   
1 2015-01-01 07:00:00          1  162.879223  402.747490   95.460525   
2 2015-01-01 08:00:00          1  170.989902  527.349825   75.237905   

   vibration  
0  45.087686  
1  43.413973  
2  34.178847  
                  datetime  machineID        volt      rotate    pressure  \
876097 2016-01-01 04:00:00        100  192.483414  447.816524   94.132837   
876098 2016-01-01 05:00:00        100  165.475310  413.771670  104.081073   
876099 2016-01-01 06:00:00        100  171.336037  496.096870   79.095538   

        vibration  
876097  48.314561  
876098  44.835259  
876099  37.845245  

--- column types ---
datetime     datetime64[us]
machineID             int64
volt                float64
rotate              float64
pressure            float64
vibration           float64
dtype: objec

In [15]:
first_look(tables["failures"], "failures")
print(tables["failures"]["datetime"].dt.hour.value_counts())

FAILURES   shape: (761, 3)

--- first and last rows ---
             datetime  machineID failure
0 2015-01-05 06:00:00          1   comp4
1 2015-03-06 06:00:00          1   comp1
2 2015-04-20 06:00:00          1   comp2
               datetime  machineID failure
758 2015-02-12 06:00:00        100   comp1
759 2015-09-10 06:00:00        100   comp1
760 2015-12-09 06:00:00        100   comp2

--- column types ---
datetime     datetime64[us]
machineID             int64
failure                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
failure      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-02 03:00:00 → 2015-12-31 06:00:00

unique machines: 98
same machine + same time rows: 42

--- value counts: failure ---
failure
comp2    259
comp1    192
comp4    179
comp3    131
Name: count, dtype: int64
datetime
6    743
3     18
Name: count, dtype: int64


In [16]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/raw")

FILES = {
    "telemetry": "PdM_telemetry.csv",
    "errors":    "PdM_errors.csv",
    "maint":     "PdM_maint.csv",
    "failures":  "PdM_failures.csv",
    "machines":  "PdM_machines.csv",
}

tables = {}
for name, file in FILES.items():
    df = pd.read_csv(DATA_DIR / file)

    if "datetime" in df.columns:
        df["datetime"] = pd.to_datetime(df["datetime"])   # only for tables that have dates
        date_info = str(df["datetime"].dtype)
    else:
        date_info = "no datetime column"                    # machines lands here

    tables[name] = df                                       # store every table, dates or not
    print(f"{name:<10} {str(df.shape):<14} {date_info}")

telemetry  (876100, 6)    datetime64[us]
errors     (3919, 3)      datetime64[us]
maint      (3286, 3)      datetime64[us]
failures   (761, 3)       datetime64[us]
machines   (100, 3)       no datetime column


In [17]:
print("Machines:", tables["failures"]["machineID"].nunique())

print(
    "Date range:",
    tables["failures"]["datetime"].min(),
    "→",
    tables["failures"]["datetime"].max()
)

print("Failures per component:")
print(tables["failures"]["failure"].value_counts())

Machines: 98
Date range: 2015-01-02 03:00:00 → 2015-12-31 06:00:00
Failures per component:
failure
comp2    259
comp1    192
comp4    179
comp3    131
Name: count, dtype: int64


In [18]:
first_look(tables["errors"], "errors")
print(tables["errors"]["datetime"].dt.hour.value_counts().head())

ERRORS   shape: (3919, 3)

--- first and last rows ---
             datetime  machineID errorID
0 2015-01-03 07:00:00          1  error1
1 2015-01-03 20:00:00          1  error3
2 2015-01-04 06:00:00          1  error5
                datetime  machineID errorID
3916 2015-12-08 06:00:00        100  error2
3917 2015-12-08 06:00:00        100  error3
3918 2015-12-22 03:00:00        100  error3

--- column types ---
datetime     datetime64[us]
machineID             int64
errorID                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
errorID      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-01 06:00:00 → 2016-01-01 05:00:00

unique machines: 100
same machine + same time rows: 303

--- value counts: errorID ---
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
datetime
6     1122
2      142
8      135
1      132
13     130
Name: count, dtype: int64


### Failures

Failures: one row = one component failure event  
Rows: 761 | Machines: 98 | Date range: 2015-01-02 03:00:00 → 2015-12-31 06:00:00  
Failures per component: comp1 192, comp2 259, comp3 131, comp4 179  
Missing values: 0 | Duplicate rows: 0  
Hour of day: most failures are logged at 6 AM

In [19]:
first_look(tables["errors"], "errors")
print(tables["errors"]["datetime"].dt.hour.value_counts().head())

ERRORS   shape: (3919, 3)

--- first and last rows ---
             datetime  machineID errorID
0 2015-01-03 07:00:00          1  error1
1 2015-01-03 20:00:00          1  error3
2 2015-01-04 06:00:00          1  error5
                datetime  machineID errorID
3916 2015-12-08 06:00:00        100  error2
3917 2015-12-08 06:00:00        100  error3
3918 2015-12-22 03:00:00        100  error3

--- column types ---
datetime     datetime64[us]
machineID             int64
errorID                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
errorID      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-01 06:00:00 → 2016-01-01 05:00:00

unique machines: 100
same machine + same time rows: 303

--- value counts: errorID ---
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
datetime
6     1122
2      142
8      135
1      132
13     130
Name: count, dtype: int64


In [20]:
first_look(tables["errors"], "errors")
print(tables["errors"]["datetime"].dt.hour.value_counts().head())

ERRORS   shape: (3919, 3)

--- first and last rows ---
             datetime  machineID errorID
0 2015-01-03 07:00:00          1  error1
1 2015-01-03 20:00:00          1  error3
2 2015-01-04 06:00:00          1  error5
                datetime  machineID errorID
3916 2015-12-08 06:00:00        100  error2
3917 2015-12-08 06:00:00        100  error3
3918 2015-12-22 03:00:00        100  error3

--- column types ---
datetime     datetime64[us]
machineID             int64
errorID                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
errorID      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-01 06:00:00 → 2016-01-01 05:00:00

unique machines: 100
same machine + same time rows: 303

--- value counts: errorID ---
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
datetime
6     1122
2      142
8      135
1      132
13     130
Name: count, dtype: int64


In [21]:
print("Machines:", tables["errors"]["machineID"].nunique())

print(
    "Date range:",
    tables["errors"]["datetime"].min(),
    "→",
    tables["errors"]["datetime"].max()
)

print("Error types:")
print(tables["errors"]["errorID"].value_counts())

print("Number of hours:", tables["errors"]["datetime"].dt.hour.nunique())

Machines: 100
Date range: 2015-01-01 06:00:00 → 2016-01-01 05:00:00
Error types:
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
Number of hours: 24


In [22]:
first_look(tables["errors"], "errors")
print(tables["errors"]["datetime"].dt.hour.value_counts().head())

ERRORS   shape: (3919, 3)

--- first and last rows ---
             datetime  machineID errorID
0 2015-01-03 07:00:00          1  error1
1 2015-01-03 20:00:00          1  error3
2 2015-01-04 06:00:00          1  error5
                datetime  machineID errorID
3916 2015-12-08 06:00:00        100  error2
3917 2015-12-08 06:00:00        100  error3
3918 2015-12-22 03:00:00        100  error3

--- column types ---
datetime     datetime64[us]
machineID             int64
errorID                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
errorID      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-01 06:00:00 → 2016-01-01 05:00:00

unique machines: 100
same machine + same time rows: 303

--- value counts: errorID ---
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
datetime
6     1122
2      142
8      135
1      132
13     130
Name: count, dtype: int64


Errors: one row = one error event  
Rows: 3919 | Machines: 100 | Date range: 2015-01-01 06:00:00 → 2016-01-01 05:00:00  
Error types: error1 1010, error2 988, error3 838, error4 727, error5 356  
Missing values: 0 | Duplicate rows: 0  
Are errors logged at one hour, or all through the day? All through the day (24 distinct hours)

In [23]:
first_look(tables["maint"], "maint")
print(tables["maint"]["datetime"].dt.hour.value_counts())

MAINT   shape: (3286, 3)

--- first and last rows ---
             datetime  machineID   comp
0 2014-06-01 06:00:00          1  comp2
1 2014-07-16 06:00:00          1  comp4
2 2014-07-31 06:00:00          1  comp3
                datetime  machineID   comp
3283 2015-11-09 06:00:00        100  comp4
3284 2015-12-09 06:00:00        100  comp2
3285 2015-12-24 06:00:00        100  comp2

--- column types ---
datetime     datetime64[us]
machineID             int64
comp                    str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
comp         0
dtype: int64
duplicate rows: 0

--- time range ---
2014-06-01 06:00:00 → 2016-01-01 06:00:00

unique machines: 100
same machine + same time rows: 758

--- value counts: comp ---
comp
comp2    863
comp4    811
comp3    808
comp1    804
Name: count, dtype: int64
datetime
6    3286
Name: count, dtype: int64


In [24]:
print("Machines:", tables["maint"]["machineID"].nunique())

print(
    "Date range:",
    tables["maint"]["datetime"].min(),
    "→",
    tables["maint"]["datetime"].max()
)

print("Replacements per component:")
print(tables["maint"]["comp"].value_counts())

Machines: 100
Date range: 2014-06-01 06:00:00 → 2016-01-01 06:00:00
Replacements per component:
comp
comp2    863
comp4    811
comp3    808
comp1    804
Name: count, dtype: int64


Maint: one row = one maintenance/replacement event  
Rows: 3286 | Machines: 100 | Date range: 2014-06-01 06:00:00 → 2016-01-01 06:00:00  
Replacements per component: comp1 804, comp2 863, comp3 808, comp4 811  
Missing values: 0 | Duplicate rows: 0  
Maint has more rows than failures. Why might that be? Maintenance can be preventive/scheduled, so a component may be replaced before it fails.

In [25]:
first_look(tables["machines"], "machines")

MACHINES   shape: (100, 3)

--- first and last rows ---
   machineID   model  age
0          1  model3   18
1          2  model4    7
2          3  model3    8
    machineID   model  age
97         98  model2   20
98         99  model1   14
99        100  model4    5

--- column types ---
machineID    int64
model          str
age          int64
dtype: object

--- missing values per column ---
machineID    0
model        0
age          0
dtype: int64
duplicate rows: 0

unique machines: 100

--- value counts: model ---
model
model3    35
model4    32
model2    17
model1    16
Name: count, dtype: int64

--- number summary ---
          age
count  100.00
mean    11.33
std      5.86
min      0.00
25%      6.75
50%     12.00
75%     16.00
max     20.00


In [26]:
print(tables["machines"]["model"].value_counts())
print("Min age:", tables["machines"]["age"].min())


model
model3    35
model4    32
model2    17
model1    16
Name: count, dtype: int64
Min age: 0


Machines: one row = one machine
Rows: 100 | Models: 4 (counts: model1 16, model2 17, model3 35, model4 32)
Age: youngest 0, oldest 20, median 12
Missing values: 0 | Duplicate rows: 0

In [27]:
for name, df in tables.items():
    if "datetime" in df.columns:
        print(f"{name:<10} {df['datetime'].min()} → {df['datetime'].max()}   "
              f"machines: {df['machineID'].nunique()}")

telemetry  2015-01-01 06:00:00 → 2016-01-01 06:00:00   machines: 100
errors     2015-01-01 06:00:00 → 2016-01-01 05:00:00   machines: 100
maint      2014-06-01 06:00:00 → 2016-01-01 06:00:00   machines: 100
failures   2015-01-02 03:00:00 → 2015-12-31 06:00:00   machines: 98


In [28]:
print("Jupyter is working")

Jupyter is working


In [29]:
print(tables["failures"].shape)

(761, 3)


In [30]:
fail = tables["failures"]

print("Hour of day:")
print(fail["datetime"].dt.hour.value_counts())

print("\nMachines that ever failed:", fail["machineID"].nunique())

print("\nFailures per machine:")
print(fail.groupby("machineID").size().describe())

Hour of day:
datetime
6    743
3     18
Name: count, dtype: int64

Machines that ever failed: 98

Failures per machine:
count    98.000000
mean      7.765306
std       3.463509
min       2.000000
25%       5.000000
50%       7.000000
75%      10.000000
max      19.000000
dtype: float64


In [31]:
id_cols = ["machineID", "datetime"]

same_time = fail.duplicated(id_cols, keep=False)

print("Rows sharing machine + same time:", same_time.sum())
print("Unique machine + time combinations:", fail[id_cols].drop_duplicates().shape[0])

Rows sharing machine + same time: 84
Unique machine + time combinations: 719


Failures: one row = one recorded component failure event
761 rows | Machines that failed: 98 | Never failed: 2
Per component: comp1 192, comp2 259, comp3 131, comp4 179
Same machine + same time rows: 84 → meaning for y: multiple components can fail at the same time, so an initial target can be any failure (yes/no)
Failures per machine: min 2, median 7, max 19
Most failures logged at hour: 6

In [32]:
first_look(tables["errors"], "errors")
print(tables["errors"]["datetime"].dt.hour.value_counts().head())

ERRORS   shape: (3919, 3)

--- first and last rows ---
             datetime  machineID errorID
0 2015-01-03 07:00:00          1  error1
1 2015-01-03 20:00:00          1  error3
2 2015-01-04 06:00:00          1  error5
                datetime  machineID errorID
3916 2015-12-08 06:00:00        100  error2
3917 2015-12-08 06:00:00        100  error3
3918 2015-12-22 03:00:00        100  error3

--- column types ---
datetime     datetime64[us]
machineID             int64
errorID                 str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
errorID      0
dtype: int64
duplicate rows: 0

--- time range ---
2015-01-01 06:00:00 → 2016-01-01 05:00:00

unique machines: 100
same machine + same time rows: 303

--- value counts: errorID ---
errorID
error1    1010
error2     988
error3     838
error4     727
error5     356
Name: count, dtype: int64
datetime
6     1122
2      142
8      135
1      132
13     130
Name: count, dtype: int64


Errors: one row = one error event

3,919 rows | Machines: 100 | 2015-01-01 06:00:00 → 2016-01-01 05:00:00

Error types: error1 1010, error2 988, error3 838, error4 727, error5 356

Logged at one hour or all day? All day (24 distinct hours)

In [33]:
first_look(tables["maint"], "maint")
print(tables["maint"]["datetime"].dt.hour.value_counts())

# How many failures also appear as a maintenance record (same machine, time, component)?
match = tables["maint"].merge(
    tables["failures"],
    left_on=["machineID", "datetime", "comp"],
    right_on=["machineID", "datetime", "failure"],
    how="inner")
print("Failures that also appear in maint:", len(match), "of", len(tables["failures"]))

MAINT   shape: (3286, 3)

--- first and last rows ---
             datetime  machineID   comp
0 2014-06-01 06:00:00          1  comp2
1 2014-07-16 06:00:00          1  comp4
2 2014-07-31 06:00:00          1  comp3
                datetime  machineID   comp
3283 2015-11-09 06:00:00        100  comp4
3284 2015-12-09 06:00:00        100  comp2
3285 2015-12-24 06:00:00        100  comp2

--- column types ---
datetime     datetime64[us]
machineID             int64
comp                    str
dtype: object

--- missing values per column ---
datetime     0
machineID    0
comp         0
dtype: int64
duplicate rows: 0

--- time range ---
2014-06-01 06:00:00 → 2016-01-01 06:00:00

unique machines: 100
same machine + same time rows: 758

--- value counts: comp ---
comp
comp2    863
comp4    811
comp3    808
comp1    804
Name: count, dtype: int64
datetime
6    3286
Name: count, dtype: int64
Failures that also appear in maint: 743 of 761


Maint: one row = one maintenance/replacement event

3,286 rows | Machines: 100 | 2014-06-01 → 2016-01-01

Replacements per component: comp1 804, comp2 863, comp3 808, comp4 811

Failures that also appear in maint: 743 of 761

So maint contains two kinds of replacement: preventive and failure-related

→ that's why maint can't be the label: a maintenance/replacement event does not necessarily mean that a failure occurred; some replacements are preventive.

In [34]:
first_look(tables["machines"], "machines")
print(tables["machines"]["age"].min(), tables["machines"]["age"].max())

MACHINES   shape: (100, 3)

--- first and last rows ---
   machineID   model  age
0          1  model3   18
1          2  model4    7
2          3  model3    8
    machineID   model  age
97         98  model2   20
98         99  model1   14
99        100  model4    5

--- column types ---
machineID    int64
model          str
age          int64
dtype: object

--- missing values per column ---
machineID    0
model        0
age          0
dtype: int64
duplicate rows: 0

unique machines: 100

--- value counts: model ---
model
model3    35
model4    32
model2    17
model1    16
Name: count, dtype: int64

--- number summary ---
          age
count  100.00
mean    11.33
std      5.86
min      0.00
25%      6.75
50%     12.00
75%     16.00
max     20.00
0 20


Machines: one row = one machine

Models: 4 (counts: model1 16, model2 17, model3 35, model4 32) | Age: 0 to 20 years, median 12

In [35]:
for name, df in tables.items():
    if "datetime" in df.columns:
        print(f"{name:<10} {df['datetime'].min()} → {df['datetime'].max()}   "
              f"machines: {df['machineID'].nunique()}")

telemetry  2015-01-01 06:00:00 → 2016-01-01 06:00:00   machines: 100
errors     2015-01-01 06:00:00 → 2016-01-01 05:00:00   machines: 100
maint      2014-06-01 06:00:00 → 2016-01-01 06:00:00   machines: 100
failures   2015-01-02 03:00:00 → 2015-12-31 06:00:00   machines: 98
